1-Draft/RFC0000-Chain-Operators.md

```diff
@@ -0,0 +1,611 @@
+of the precedence of the background operator.
+
+The example:
+
+```powershell
+cmd1 && cmd2 &
+```
+
+Could either be `[[cmd1 && cmd2] &]` or `[cmd1 && [cmd2 &]]`.
+
+Having chain operators apply to statements would mean the second case
+is the correct one, unless changes are made to the background operator precedence (which would also allow `if ($condition) { cmd1 } &)` for example).
+
+The use of this is possibly less than the use of being able to background
+an entire pipeline chain; `cmd1 & && cmd2` would be equivalent to
+`cmd1 &; cmd2` since a backgrounded pipeline
+will never fail in the invoking context.
+
+Such a change to the background operator precedence
+would cause ambiguity with uses like `return cmd1 &` and `throw cmd1 &`,
+where the background operator would currently apply to the pipeline under the keyword.
+To not break PowerShell's existing semantics,
+special behaviour would need to be defined for `return $expr &`.
+
+#### Reasons against
+
+- Pipelines have an established concept of "success" compared to statements
+- Background operators become less useful with respect to chains unless their
+  syntax is changed in a significant way
+
+### Allowing control flow statements at the end of chains
+
+A compromise to the above is to only allow "control flow statements"
+at the end of chains.
+
+For example:
+
+```powershell
+cmd1 || throw "cmd1 failed"
+```
+
+```powershell
+function Invoke-Command
+{
+    cmd1 && return
+
+    cmd2
+}
+```
+
+```powershell
+foreach ($v in 1..100)
+{
+    cmd1 $v && break
+}
+```
+
+These would only be allowed *at the end* of chains because:
+
+- Control flow makes any invocation after the statement unreachable
+- `return` and `throw` allow pipelines as subordinate expressions,
+   meaning it would be grammatically impossible to use those statements mid-chain.
+   See below for more details.
+
+#### Reasons against
+
+- `throw` and `return` currently allow a subordinate pipeline:
+
+   ```powershell
+   throw "Bad"
+   ```
+
+   ```powershell
+   throw 1,2,3 | Write-Output
+   ```
+
+   ```powershell
+   throw "Useless" & # Throws a background job...
+   ```
+
+   If we keep the principle that anywhere a pipeline is allowed,
+   a pipeline chain is now allowed, then constructions like this are possible:
+
+   ```powershell
+   cmd1 || throw "Bad" && cmd3
+   ```
+
+   However the grouping is this:
+
+   ```none
+   [cmd1 || throw ["Bad" && cmd3]]
+   ```
+
+   This could lead to confusing semantic corner cases
+
+- The compromise nature of this approach means the PowerShell Language
+  becomes more complicated and arguably less consistent,
+  both conceptually and in terms of maintenance.
+
+### Different evaluations of command "success"
+
+The current proposal is to simply use `$?` to determine chain continuation.
+
+Alternatives include:
+
+- `$LASTEXITCODE`
+- Whether errors have been written
+- A new custom semantics
+

```

Comment:
Jobs could/should be handled differently. See my brain dump above about how that might work.